### PR TITLE
Make multigrid function `fill_copy_indices` available internally

### PR DESCRIPTION
**Pull request by kalj**
I am writing a new matrix-free multigrid transfer class for GPUs, and then I need to access this function from within `mg_level_global_transfer.h`. This puts it in namespace `internal` and makes it available.

### CODE REVIEW

File: source/multigrid/mg_level_global_transfer.cc [hunk 1]
```diff
@@ -291,7 +293,7 @@ void
 MGLevelGlobalTransfer<VectorType>::fill_and_communicate_copy_indices
 (const DoFHandler<dim,spacedim> &mg_dof)
 {
-  fill_copy_indices(mg_dof, mg_constrained_dofs, copy_indices,
+  internal::fill_copy_indices(mg_dof, mg_constrained_dofs, copy_indices,
                     copy_indices_global_mine, copy_indices_level_mine);
```
Comment (bangerth) [inline source/multigrid/mg_level_global_transfer.cc]: The indentation script complains about this line and one further down. You'll need to run `./contrib/utilities/indent`.

File: source/multigrid/mg_level_global_transfer.cc [hunk 2]
```diff
@@ -70,11 +70,14 @@ namespace
 
 
 
+}
+
+namespace internal {
```
Comment (bangerth) [inline source/multigrid/mg_level_global_transfer.cc]: break the brace onto the next line

**Comment by bangerth** [comment]
I'm not too enthusiastic to export a global function. Is this function typically called from a class? Could it be made into a member function?

**Comment by kalj** [comment]
It is hidden inside internal, but I see your point. I guess we could change the name to something more qualified. Subclasses of MGLevelGlobalTransfer gets access to it through the member function `fill_and_communicate_copy_indices`, but my class doesn't derive from that one. We (me and @kronbichler ) also discussed something like `include/deal.II/lac/vector_operations_internal.h`, i.e. putting it in a separate `mg_transfer_internal.h` file which we can then include from `mg_level_global_transfer.cc` and from my corresponding file.

**Comment by kalj** [comment]
Also, I should add that there is also some common code from `MGTransferMatrixFree` that we are going to have to extract to functions somewhere.

**Comment by bangerth** [comment]
It doesn't matter to me which file it is actually declared in. 

An alternative, if you need it to be a free function, is to put it into a namespace inside `internal`, say `internal::MGTools` (or something that includes the term "transfer"). This way, at least the fully qualified name indicates where the function belongs.

**Comment by kalj** [comment]
Right. A namespace sounds like a good idea. I'll think of something and update this.

**Comment by kronbichler** [comment]
@bangerth 
> An alternative, if you need it to be a free function [...]

I think that a free function is the most appropriate way of doing things here because different classes (at the same level of the heritage hierarchy) use the function and there is no common data or other things that support adding another base class.

> It doesn't matter to me which file it is actually declared in.

In that case we should put the functions in `include/deal.II/lac/vector_operations_internal.h` into a namespace `dealii::internal::VectorOperations` rather than `dealii::internal` before the next release. We did not release that re-factored code before (it comes from #2577). It follows the exact same pattern. 
Anyway, I agree that adding another namespace that describes the category the internal function belongs to is useful. If you agree, please open an issue so this does not get lost.

**Comment by bangerth** [comment]
OK, great. I assume with "you" in the last sentence you meant @kalj ? I concur with this interpretation of "you" ;-)

**Comment by kalj** [comment]
How about `internal::MGTransfer` or simply `internal::Multigrid`? I don't quite like MGTools because I think the infrastructure we are talking about here do not really deserve to be call tools.

Regarding the free function or not, I agree that that's the natural way for this one. But for the setup code, it is more tricky since there is so much data that is being affected. That requires a separate discussion.

**Comment by kronbichler** [comment]
I would be fine with `internal::MGTransfer`.

**Comment by bangerth** [comment]
I'm happy with any choice of namespace name you choose.

**Comment by kalj** [comment]
I've updated it now with a new file `mg_transfer_internal.h` and a new namespace `internal::MGTransfer`, which we will both reuse soon for exporting a setup function in the same way.

**Comment by kalj** [comment]
Hm, actually compiling your code before a PR is not a bad idea... Now it works, so please check if this is ready for merge.

File: include/deal.II/multigrid/mg_transfer_internal.h [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace internal
+{
+  namespace MGTransfer
+  {
+    /**
+     * Internal data structure that is used in the MPI communication in fill_and_communicate_copy_indices().
+     * It represents an entry in the copy_indices* map, that associates a level dof index with a global dof index.
```
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_internal.h]: Please break lines at 80 characters.

Comment (kalj) [inline include/deal.II/multigrid/mg_transfer_internal.h]: Fixed in new commit.

**Comment by kronbichler** [comment]
This looks good to me but I let @bangerth have a final look since he requested changes.

**Comment by bangerth** [comment]
It looks good to me, but it doesn't compile right now:
  http://www.math.clemson.edu/~heister/cib/dealii-tester-alpha/5400ae58f0f2357afdc19e5274f43ab4724f1a8e/log-clang

The `.h` file is missing some `#include`s to make it self-contained. But that of course raises another question: does the entire function have to be in the `.h` file? I think I would prefer it if the function was just *declared* in the `.h` file, then implemented in a `.cc` file, and explicitly instantiated. This way, if you want to call it, you don't have to re-compile the whole thing, nor transitively `#include` everything that is necessary to compile it.

**Comment by kalj** [comment]
Missing includes are added now (pretty ingenious test btw).

About putting implementations in a `.cc` file, I don't see any reason against that, except maybe that we get another source file. If no one else objects, I'll go down that path.

**Comment by bangerth** [comment]
Please go ahead with the .cc plan.

**Comment by kalj** [comment]
Okay, let's see if it looks good now.

**Comment by bangerth** [comment]
There are a lot of test failures: http://www.math.clemson.edu/~heister/cib/dealii-tester-alpha/8fb9853f4a91ac0ab104d63b43c717a548bfbe4e/log-clang

I think this has to do with the fact that you access things like `DoFHandler<3,1>`, which clearly makes no sense. You need to guard instantiations with `deal_II_dimension <= deal_II_space_dimension`. See several of the other `.inst.in` files.

File: source/multigrid/mg_transfer_internal.inst.in [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+
+for (deal_II_dimension : DIMENSIONS;
+        deal_II_space_dimension: SPACE_DIMENSIONS)
```
Comment (bangerth) [inline source/multigrid/mg_transfer_internal.inst.in]: need the guard via `#if` in the body of this loop, see my earlier comment

**Comment by kalj** [comment]
Ah yes. Fixed now, but I still don't understand how that linking error could be raised from that test. I reckon the test does not try to call those badly configured types?

Speaking of this, just out of curiosity,  couldn't one have a (compile time) check for correct template parameters, presumably inside `DoFHandler` in this case?

**Comment by bangerth** [comment]
It's nothing in the test. It's just that in that one .cc file, you're referencing symbols that the linker later can't find because they have not been instantiated.

I think that in C++11, you can have `static_assert` statements at class level. You could try if this actually works, and if so submit this as a separate pull request:
```
  template <int dim, int spacedim>
  class DoFHandler {
    static_assert (dim<=spacedim);
    ....
```

**Comment by bangerth** [comment]
This works:
```
  static_assert (dim<=spacedim,
                 "The dimension <dim> of a DoFHandler must be less than or "
                 "equal to the space dimension <spacedim> in which it lives.");
```

But all this is tangential to your patch. Let's stay focussed on that.

**Comment by bangerth** [comment]
OK to merge if the tester is happy with it.

**Pull request merged**